src/Catalog/Icons/IconProcessor.cs

```diff
@@ -0,0 +1,274 @@
+                    totalBytesRead += bytesRead;
+                    await ms.WriteAsync(buffer, 0, bytesRead, cancellationToken);
+                } while (bytesRead > 0 && totalBytesRead < maxBytes + 1);
+
+                if (totalBytesRead > maxBytes)
+                {
+                    _logger.LogInformation("Source data too long, discarding.");
+                    return null;
+                }
+
+                return ms.ToArray();
+            }
+        }
+
+        /// <summary>
+        /// The PNG file header bytes. All PNG files are expected to have those at the beginning of the file.
+        /// </summary>
+        /// <remarks>
+        /// https://www.w3.org/TR/PNG/#5PNG-file-signature
+        /// </remarks>
+        private static readonly byte[] PngHeader = new byte[] { 0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A };
+
+        /// <summary>
+        /// The JPG file header bytes.
+        /// </summary>
+        /// <remarks>
+        /// Technically, JPEG start with two byte SOI (start of image) segment: FFD8, followed by several other segments or fill bytes.
+        /// All of the segments start with FF, and fill bytes are FF, so we check the first 3 bytes instead of the first two.
+        /// https://www.w3.org/Graphics/JPEG/itu-t81.pdf "B.1.1.2 Markers"
+        /// </remarks>
+        private static readonly byte[] JpegHeader = new byte[] { 0xFF, 0xD8, 0xFF };
+
+        /// <summary>
+        /// The GIF87a file header.
+        /// </summary>
+        /// <remarks>
+        /// https://www.w3.org/Graphics/GIF/spec-gif87.txt
+        /// </remarks>
+        private static readonly byte[] Gif87aHeader = new byte[] { 0x47, 0x49, 0x46, 0x38, 0x37, 0x61 };
+
+        /// <summary>
+        /// The GIF89a file header.
+        /// </summary>
+        /// <remarks>
+        /// https://www.w3.org/Graphics/GIF/spec-gif89a.txt
+        /// </remarks>
+        private static readonly byte[] Gif89aHeader = new byte[] { 0x47, 0x49, 0x46, 0x38, 0x39, 0x61 };
+
+        /// <summary>
+        /// The .ico file "header".
+        /// </summary>
+        /// <remarks>
+        /// This is the first 4 bytes of the ICONDIR structure expected for .ico files
+        /// https://docs.microsoft.com/en-us/previous-versions/ms997538(v=msdn.10)
+        /// </remarks>
+        private static readonly byte[] IcoHeader = new byte[] { 0x00, 0x00, 0x01, 0x00 };
+
+        private static string DetermineContentType(byte[] imageData, bool onlyGallerySupported)
+        {
+            // checks are ordered by format popularity among external icons for existing packages
+
+            if (ArrayStartsWith(imageData, PngHeader))
+            {
+                return "image/png";
+            }
+
+            if (ArrayStartsWith(imageData, JpegHeader))
+            {
+                return "image/jpeg";
+            }
+
+            if (onlyGallerySupported)
```

Comment:
>onlyGallerySupported [](start = 16, length = 20)

Does this become application/octet-stream eventually? Is this consistent with catalog2dnx

Reply:
I want to pull this code to catalog2dnx eventually, so we have proper content type everywhere.